### PR TITLE
Qabansi aut 6598 sdk ocean ecs

### PR DESCRIPTION
**Pull request by chaplabansi**
Implementation of Ocean Ecs CRUD Operations

### CODE REVIEW

File: src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/OceanEcsCluster.java [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+    private String                                 id;
+    private String                                 name;
+    private String                                 region;
+    private String                                 clusterName;
+    private ClusterAutoScalerConfiguration         autoScaler;
+    private ClusterCapacityConfiguration           capacity;
+    private ClusterComputeConfiguration            compute;
+    private ClusterSchedulingConfiguration         scheduling;
+    private ClusterStrategyConfiguration           strategy;
+    private Date                                   createdAt;
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/OceanEcsCluster.java:25]: Do we need 'createdAt' and 'updatedAt' fields.

File: src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/OceanEcsCluster.java [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+
+/**
+ * Created by Bansi Chapla on 20/07/2021.
+ */
+
+public class OceanEcsCluster {
+
+    @JsonIgnore
+    private Set<String>                            isSet;
+    private String                                 id;
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/OceanEcsCluster.java:16]: Is this field needed.

File: src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/ClusterStrategyConfiguration.java [hunk 3]
```diff
@@ -13,6 +12,7 @@
     private Boolean     utilizeReservedInstances;
     private Integer     drainingTimeout;
     private Integer     spotPercentage;
+    private Boolean     utilizeCommitments;
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/ClusterStrategyConfiguration.java:15]: fallbackToOd parameter is missing.
utilizeCommitments parameter is not present in the strategy object.

File: src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/ClusterOptimizeImageConfiguration.java [hunk 4]
```diff
@@ -7,13 +7,12 @@
 import java.util.Set;
 
 public class ClusterOptimizeImageConfiguration {
-
+    @JsonIgnore
     private Set<String>  isSet;
     private Boolean      shouldOptimizeEcsAmi;
     private String       performAt;
     private List<String> timeWindows;
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/ClusterOptimizeImageConfiguration.java:14]: Please reconfirm if it should be list of strings.

File: src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/ClusterLaunchSpecification.java [hunk 5]
```diff
@@ -24,6 +24,8 @@
     private List<LaunchSpecTagsSpecification>         tags;
     private String                                    userData;
     private String                                    id;
+    private String                                    keyPair;
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/ClusterLaunchSpecification.java:27]: Didn't find these two parameters in the LaunchSpecification.

File: src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/ClusterInstanceTypes.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
+package com.spotinst.sdkjava.model.bl.ocean.ecs;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public class ClusterInstanceTypes {
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/ClusterInstanceTypes.java:9]: Didn't find this parameter at cluster level.

File: src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/ClusterComputeConfiguration.java [hunk 7]
```diff
@@ -11,8 +10,9 @@
     @JsonIgnore
     private Set<String>                       isSet;
     private List<String>                      subnetIds;
-    private ClusterOptimizeImageConfiguration optimizeImage;
+    private ClusterOptimizeImageConfiguration optimizeImages;
     private ClusterLaunchSpecification        launchSpecification;
+    private ClusterInstanceTypes              instanceTypes;
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/ClusterComputeConfiguration.java:15]: Didn't find this parameter in the API specification.

File: src/main/java/com/spotinst/sdkjava/model/api/ocean/ecs/ApiOceanEcsCluster.java [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+    private String                                 id;
+    private String                                 name;
+    private String                                 region;
+    private String                                 clusterName;
+    private ApiClusterAutoScalerConfiguration      autoScaler;
+    private ApiClusterCapacityConfiguration        capacity;
+    private ApiClusterComputeConfiguration         compute;
+    private ApiClusterSchedulingConfiguration      scheduling;
+    private ApiClusterStrategyConfiguration        strategy;
+    private Date                                   createdAt;
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/api/ocean/ecs/ApiOceanEcsCluster.java:34]: Is these Date parameters needed?

File: src/main/java/com/spotinst/sdkjava/model/SpotOceanEcsClusterClient.java [hunk 9]
```diff
@@ -143,4 +155,105 @@ public ClusterLaunchSpecification getLaunchSpec(OceanEcsClusterLaunchSpecRequest
 
         return retVal;
     }
+
+    public  OceanEcsCluster createOceanEcsCluster(OceanEcsClusterRequest oceanEcsClusterCreationRequest)
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/SpotOceanEcsClusterClient.java:159]: Provide line breaks before and after if conditions.

**Review by anuragsharma-123 — CHANGES_REQUESTED** [review]
Please review the comments and make the appropriate changes.

File: src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/OceanEcsCluster.java (hunk 2 above)
Comment (chaplabansi) [reply src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/OceanEcsCluster.java:16]: Yes, Id is needed. we are setting this id with Id we get in cluster creation and then using same id for update, delete operation.

File: src/main/java/com/spotinst/sdkjava/model/SpotOceanEcsClusterClient.java (hunk 9 above)
Comment (chaplabansi) [reply src/main/java/com/spotinst/sdkjava/model/SpotOceanEcsClusterClient.java:159]: Incorporated.

File: src/main/java/com/spotinst/sdkjava/model/api/ocean/ecs/ApiOceanEcsCluster.java (hunk 8 above)
Comment (chaplabansi) [reply src/main/java/com/spotinst/sdkjava/model/api/ocean/ecs/ApiOceanEcsCluster.java:34]: Removed Date Parameters.

File: src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/ClusterOptimizeImageConfiguration.java (hunk 4 above)
Comment (chaplabansi) [reply src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/ClusterOptimizeImageConfiguration.java:14]: Checked for Elastigroup and Ocean Kubernetes Cluster - timewindows parameter is list of String.

File: src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/ClusterLaunchSpecification.java (hunk 5 above)
Comment (chaplabansi) [reply src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/ClusterLaunchSpecification.java:27]: Parameters keyPair and Instance Types are available in JSON when we check 'view configuration' for Cluster on UI. So Added it. 
So As Discussed, Currently aligning with postman and removing those parameters.

File: src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/ClusterComputeConfiguration.java (hunk 7 above)
Comment (chaplabansi) [reply src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/ClusterComputeConfiguration.java:15]: Parameters keyPair and Instance Types are available in JSON when we check 'view configuration' for Cluster on UI. So Added it. 
So As Discussed, Currently aligning with postman and removing those parameters.

File: src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/ClusterInstanceTypes.java (hunk 6 above)
Comment (chaplabansi) [reply src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/ClusterInstanceTypes.java:9]: Parameters keyPair and Instance Types are available in JSON when we check 'view configuration' for Cluster on UI. So Added it. 
So As Discussed, Currently aligning with postman and removing those parameters.

File: src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/OceanEcsCluster.java (hunk 1 above)
Comment (chaplabansi) [reply src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/OceanEcsCluster.java:25]: Not needed. Tried to Aligned with Elastigroup creation Example. 
Removed CreatedAt and UpdatedAt fields.

File: src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/OceanEcsCluster.java (hunk 2 above)
Comment (chaplabansi) [reply src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/OceanEcsCluster.java:16]: Yes, To get cluster id this field is required.

File: src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/ClusterStrategyConfiguration.java (hunk 3 above)
Comment (chaplabansi) [reply src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/ClusterStrategyConfiguration.java:15]: fallbackToOd parameter is already there, here is the reference: 
https://github.com/spotinst/spotinst-sdk-java/blob/qabansi_aut_6598_sdk_ocean_ecs/src/main/java/com/spotinst/sdkjava/model/bl/ocean/ecs/ClusterStrategyConfiguration.java
utilizeCommitments parameter was present in the config JSON we get on UI. Removed now, to align with Postman Request parameters.

**Comment by chaplabansi** [comment]
Incorporated Comments.

File: src/main/java/com/spotinst/sdkjava/model/repo/ocean/ecs/SpotOceanEcsClusterRepo.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+        return retVal;
+    }
+
+    @Override
+    public RepoGenericResponse<Boolean> delete(String identifier, String authToken, String account) {
+        RepoGenericResponse<Boolean> retVal;
+
+        try {
+            Boolean delete = SpotOceanEcsClusterService.deleteEcsCluster(identifier, authToken, account);
```
Comment (zivmessing) [inline src/main/java/com/spotinst/sdkjava/model/repo/ocean/ecs/SpotOceanEcsClusterRepo.java:77]: rename to isDeleted (boolean should start with a question like is/did etc

File: src/main/java/com/spotinst/sdkjava/model/repo/ocean/ecs/SpotOceanEcsClusterRepo.java [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+
+    @Override
+    public RepoGenericResponse<Boolean> update(String clusterId, OceanEcsCluster clusterUpdate, String authToken,
+                                               String account) {
+        RepoGenericResponse<Boolean> retVal;
+
+        ApiOceanEcsCluster apiOceanEcsCluster = OceanEcsConverter.toDal(clusterUpdate);
+
+        try {
+            Boolean success = SpotOceanEcsClusterService
```
Comment (zivmessing) [inline src/main/java/com/spotinst/sdkjava/model/repo/ocean/ecs/SpotOceanEcsClusterRepo.java:44]: rename to isSucceeded (boolean should start with a question : is/did/etc

File: src/main/java/com/spotinst/sdkjava/model/ISpotOceanEcsClusterRepo.java [hunk 12]
```diff
@@ -0,0 +1,7 @@
+package com.spotinst.sdkjava.model;
```
Comment (zivmessing) [inline src/main/java/com/spotinst/sdkjava/model/ISpotOceanEcsClusterRepo.java:1]: General comment:
1. in your tests try to create entity and then retrieve it in debug and compare between the object that send to the create method and the one that retrieved
2. check the update for a several fields, Please ask Shibel after he will comback from his vacation to give you handover for the partial update mechanism (for a massive creation for tests that update entity and verify that the entity updated

File: src/main/java/com/spotinst/sdkjava/model/SpotOceanEcsClusterClient.java [hunk 13]
```diff
@@ -143,4 +155,110 @@ public ClusterLaunchSpecification getLaunchSpec(OceanEcsClusterLaunchSpecRequest
...
+
+        if (creationResponse.isRequestSucceed()) {
+            retVal = creationResponse.getValue();
+        }
+        else {
+            List<HttpError> httpExceptions = creationResponse.getHttpExceptions();
+            HttpError       httpException  = httpExceptions.get(0);
+            LOGGER.error(String.format("Error encountered while attempting to create ocean ecs cluster. Code: %s. Message: %s.",
+                                       httpException.getCode(), httpException.getMessage()));
+            throw new SpotinstHttpException(httpException.getMessage());
```
[the remark this answers is not in the document]
Comment (chaplabansi) [reply src/main/java/com/spotinst/sdkjava/model/SpotOceanEcsClusterClient.java:174]: Other APIs are also throwing code and messages both. 
ex. createK8sCluster in (SpotOceanK8sClusterClient)

File: src/main/java/com/spotinst/sdkjava/model/api/ocean/ecs/ApiClusterAutoScalerConfiguration.java [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+        return isSet.contains("isEnabled");
+    }
+
+    @JsonIgnore
+    public boolean isCooldownSet() {
+        return isSet.contains("cooldown");
+    }
+
+    @JsonIgnore
+    public boolean isResourceLimits() {
```
[the remark this answers is not in the document]
Comment (chaplabansi) [reply src/main/java/com/spotinst/sdkjava/model/api/ocean/ecs/ApiClusterAutoScalerConfiguration.java:143]: Incorporated.

File: src/main/java/com/spotinst/sdkjava/model/repo/ocean/ecs/SpotOceanEcsClusterRepo.java (hunk 11 above)
Comment (chaplabansi) [reply src/main/java/com/spotinst/sdkjava/model/repo/ocean/ecs/SpotOceanEcsClusterRepo.java:44]: Incorporated

File: src/main/java/com/spotinst/sdkjava/model/repo/ocean/ecs/SpotOceanEcsClusterRepo.java (hunk 10 above)
Comment (chaplabansi) [reply src/main/java/com/spotinst/sdkjava/model/repo/ocean/ecs/SpotOceanEcsClusterRepo.java:77]: Incorporated.

File: src/main/java/com/spotinst/sdkjava/model/SpotOceanEcsClusterClient.java [hunk 15]
```diff
@@ -16,13 +19,22 @@
...
     //Constructor
     public SpotOceanEcsClusterClient(String authToken, String account) {
         this.authToken = authToken;
         this.account = account;
-
         setSpotinstOceanClusterRepo();
+        setSpotinstOceanEcsClusterRepo();
+    }
+
+    public ISpotOceanEcsClusterRepo getSpotOceanEcsClusterRepo() {
```
[the remark this answers is not in the document]
Comment (chaplabansi) [reply src/main/java/com/spotinst/sdkjava/model/SpotOceanEcsClusterClient.java:32]: changed to Private now.

File: src/main/java/com/spotinst/sdkjava/model/SpotOceanEcsClusterClient.java [hunk 16]
```diff
@@ -16,13 +19,22 @@
...
+        return spotOceanEcsClusterRepo; }
+
+    public void setSpotinstOceanEcsClusterRepo(){
```
[the remark this answers is not in the document]
Comment (chaplabansi) [reply src/main/java/com/spotinst/sdkjava/model/SpotOceanEcsClusterClient.java:35]: Made it private and removed setter method.  Instead used Constructor.

File: src/main/java/com/spotinst/sdkjava/model/SpotOceanEcsClusterClient.java [hunk 17]
```diff
@@ -143,4 +155,110 @@ public ClusterLaunchSpecification getLaunchSpec(OceanEcsClusterLaunchSpecRequest
...
+            throw new SpotinstHttpException(httpException.getMessage());
+        }
+
+        return retVal;
+    }
+
+    public List<OceanEcsCluster> getAllOceanEcsClusters() {
+        List<OceanEcsCluster> retVal;
+
+        RepoGenericResponse<List<OceanEcsCluster>> clusterRes = getSpotOceanEcsClusterRepo().getAll(null, authToken, account);
```
[the remark this answers is not in the document]
Comment (chaplabansi) [reply src/main/java/com/spotinst/sdkjava/model/SpotOceanEcsClusterClient.java:235]: First parameter is for Filter. we are just getting list of all cluster without any filter so it's null.

File: src/main/java/com/spotinst/sdkjava/model/ISpotOceanEcsClusterRepo.java (hunk 12 above)
Comment (chaplabansi) [reply src/main/java/com/spotinst/sdkjava/model/ISpotOceanEcsClusterRepo.java:1]: 1. Tested Object that is being send and retrieved after create method. 
2. Update - Currently Tested all possible attributes values with updated and verified it's updated.  So changed few attribute values in example file also. 
Regarding update Partial Mechanism - Yes, I will check with Shibel once he is back from vacation.

**Review by anuragsharma-123 — APPROVED** [review]

**Pull request merged**